authorization/storage.go

```diff
@@ -2,33 +2,313 @@ package authorization
+func (s *Store) autogenerateHasher(ctx context.Context, variantName string) (*AuthorizationHasher, error) {
+	// Determine which variants are present in the store.
+	tempDecoder := crypt.NewDecoder()
+	if err := influxdb2_algo.RegisterDecoder(tempDecoder); err != nil {
+		return nil, fmt.Errorf("error registering test decoders for authorization store: %w", err)
+	}
+
+	var auths []*influxdb.Authorization
+	err := s.View(ctx, func(tx kv.Tx) error {
+		as, err := s.ListAuthorizations(ctx, tx, influxdb.AuthorizationFilter{})
+		if err != nil {
+			return err
+		}
+		auths = as
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("error fetching authorization records for hash variant inventory: %w", err)
+	}
+
+	foundVariants := make(map[influxdb2_algo.Variant]struct{})
+	for _, a := range auths {
+		if a.HashedToken != "" {
+			digest, err := tempDecoder.Decode(a.HashedToken)
+			if err == nil {
+				if influxdbDigest, ok := digest.(*influxdb2_algo.Digest); ok {
+					foundVariants[influxdbDigest.Variant] = struct{}{}
+				}
+			}
+		}
+	}
+
+	var decoderVariants []influxdb2_algo.Variant
+	// Make sure we have the hasher variant we will make in there and that it is first in the list.
+	hasherVariant := influxdb2_algo.NewVariant(variantName)
+	decoderVariants = append(decoderVariants, hasherVariant)
+	delete(foundVariants, hasherVariant)
+	for variant := range foundVariants {
+		decoderVariants = append(decoderVariants, variant)
+	}
+
+	hasher, err := NewAuthorizationHasher(WithHasherVariant(hasherVariant), WithDecoderVariants(decoderVariants))
+	if err != nil {
+		return nil, fmt.Errorf("error creating authorization hasher for authorization store: %w", err)
+	}
+
+	return hasher, nil
+}
+
+// hashedTokenMigration migrates any unhashed tokens in the store to hashed tokens.
+func (s *Store) hashedTokenMigration(ctx context.Context) error {
+	if !s.useHashedTokens || s.readOnly {
+		return nil
+	}
+
+	// Figure out which authorization records need to be updated.
+	var authsNeedingUpdate []*influxdb.Authorization
+	err := s.View(ctx, func(tx kv.Tx) error {
+		s.forEachAuthorization(ctx, tx, nil, func(a *influxdb.Authorization) bool {
+			if a.HashedToken == "" {
+				if a.Token != "" {
+					authsNeedingUpdate = append(authsNeedingUpdate, a)
+				} else {
+					s.log.Warn("found authorization without any token set during hashed token migration", zap.Uint64("ID", uint64(a.ID)), zap.String("description", a.Description))
+				}
+			}
+			return true
+		})
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+
+	for batch := range slices.Chunk(authsNeedingUpdate, 100) {
+		err := s.Update(ctx, func(tx kv.Tx) error {
+			// Now update them. This really seems too simple, but s.UpdateAuthorization() is magical.
+			for _, a := range batch {
+				if _, err := s.UpdateAuthorization(ctx, tx, a.ID, a); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return fmt.Errorf("error migrating hashed tokens: %w", err)
```

Comment:
Adding some more context to the error such as ID and maybe token description? Similar to the s.log.Warn above?